src/main/java/edu/kit/kastel/sdq/artemis4j/grading/penalty/GradingConfig.java

```diff
@@ -0,0 +1,103 @@
+package edu.kit.kastel.sdq.artemis4j.grading.penalty;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import edu.kit.kastel.sdq.artemis4j.grading.ProgrammingExercise;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public final class GradingConfig {
+    private static final Logger log = LoggerFactory.getLogger(GradingConfig.class);
+    private final String shortName;
+    private final List<RatingGroup> ratingGroups;
+    private final long validExerciseId;
+
+    private GradingConfig(String shortName, List<RatingGroup> ratingGroups, long validExerciseId) {
+        this.shortName = shortName;
+        this.ratingGroups = ratingGroups;
+        this.validExerciseId = validExerciseId;
+    }
+
+    public static GradingConfig readFromString(String configString, ProgrammingExercise exercise) throws InvalidGradingConfigException {
+        ObjectMapper mapper = new ObjectMapper();
+
+        try {
+            var configDTO = mapper.readValue(configString, GradingConfigDTO.class);
+
+            // no allowed exercises means it is valid for all exercises
+            if (!configDTO.allowedExercises().isEmpty() && !configDTO.allowedExercises().contains(exercise.getId())) {
+                throw new InvalidGradingConfigException("Grading config is not valid for exercise with id " + exercise.getId());
+            }
+
+            var ratingGroups = configDTO.ratingGroups().stream().map(RatingGroup::new).toList();
+            var ratingGroupsById = ratingGroups.stream().collect(Collectors.toMap(RatingGroup::getId, Function.identity()));
+
+            for (MistakeType.MistakeTypeDTO dto : configDTO.mistakeTypes()) {
+                // TODO add back in maybe?
+                /*if (!StringUtil.matchMaybe(exercise.getShortName(), dto.enabledForExercises())) {
+                    continue;
+                }*/
+
+                new MistakeType(
```

Comment:
What happens here .. just new ?

Reply:
The constructor has side-effects, it adds itself to the RatingGroup.

Reply:
I will add a static method to make this more clear